File: array/40.py
class Solution:
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ret = []
        candidates.sort()
        self.backtrack(ret, [], candidates, target, 0)
        return ret
        
    def backtrack(self, ret, temp, candidates, value, start):
        if value < 0:
            return
        elif value == 0:
            ret.append(temp)
        else:
            for i in range(start, len(candidates)):
                if start < i and candidates[i-1] == candidates[i]:
                    continue
                else:
                    self.backtrack(ret, temp+[candidates[i]], candidates, value - candidates[i], i+1)

File: backtracking/40.cc
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        std::vector<std::vector<int>> ret;
        std::vector<int> cur;
        std::sort(candidates.begin(), candidates.end());
        dfs(target, 0, ret, cur, candidates, 0);
        return ret;
    }
    
private:
    void dfs(int target, int sum, std::vector<std::vector<int>>& ret, std::vector<int>& cur, std::vector<int>& candis, int start) {
        if (sum > target) {
            return;
        }
        if (target == sum) {
            ret.emplace_back(cur);
            return;
        }
        for (int i = start; i < candis.size(); i++) {
            if (i > start && candis[i] == candis[i-1]) {
                continue;
            }
            cur.emplace_back(candis[i]);
            dfs(target, sum+candis[i], ret, cur, candis, i+1);
            cur.pop_back();
        }
    }
};
